// @tgwf/co2/dist/cjs/constants/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  DATACENTER_ENERGY: () => DATACENTER_ENERGY,
  END_USER_DEVICE_ENERGY: () => END_USER_DEVICE_ENERGY,
  FIRST_TIME_VIEWING_PERCENTAGE: () => FIRST_TIME_VIEWING_PERCENTAGE,
  GLOBAL_GRID_INTENSITY: () => GLOBAL_GRID_INTENSITY,
  KWH_PER_GB: () => KWH_PER_GB,
  NETWORK_ENERGY: () => NETWORK_ENERGY,
  PERCENTAGE_OF_DATA_LOADED_ON_SUBSEQUENT_LOAD: () => PERCENTAGE_OF_DATA_LOADED_ON_SUBSEQUENT_LOAD,
  PRODUCTION_ENERGY: () => PRODUCTION_ENERGY,
  RENEWABLES_GRID_INTENSITY: () => RENEWABLES_GRID_INTENSITY,
  RETURNING_VISITOR_PERCENTAGE: () => RETURNING_VISITOR_PERCENTAGE,
  SWDMV3_RATINGS: () => SWDMV3_RATINGS,
  SWDMV4_RATINGS: () => SWDMV4_RATINGS,
  SWDV4: () => SWDV4,
  fileSize: () => import_file_size.default
});
module.exports = __toCommonJS(constants_exports);
var import_file_size = __toESM(require("./file-size.js"));
var import_average_intensities_min = __toESM(require("../data/average-intensities.min.js"));
const KWH_PER_GB = 0.81;
const END_USER_DEVICE_ENERGY = 0.52;
const NETWORK_ENERGY = 0.14;
const DATACENTER_ENERGY = 0.15;
const PRODUCTION_ENERGY = 0.19;
const GLOBAL_GRID_INTENSITY = import_average_intensities_min.default.data["WORLD"];
const RENEWABLES_GRID_INTENSITY = 50;
const FIRST_TIME_VIEWING_PERCENTAGE = 0.75;
const RETURNING_VISITOR_PERCENTAGE = 0.25;
const PERCENTAGE_OF_DATA_LOADED_ON_SUBSEQUENT_LOAD = 0.02;
const SWDV4 = {
  OPERATIONAL_KWH_PER_GB_DATACENTER: 0.055,
  OPERATIONAL_KWH_PER_GB_NETWORK: 0.059,
  OPERATIONAL_KWH_PER_GB_DEVICE: 0.08,
  EMBODIED_KWH_PER_GB_DATACENTER: 0.012,
  EMBODIED_KWH_PER_GB_NETWORK: 0.013,
  EMBODIED_KWH_PER_GB_DEVICE: 0.081,
  GLOBAL_GRID_INTENSITY: 494
};
const SWDMV3_RATINGS = {
  FIFTH_PERCENTILE: 0.095,
  TENTH_PERCENTILE: 0.186,
  TWENTIETH_PERCENTILE: 0.341,
  THIRTIETH_PERCENTILE: 0.493,
  FORTIETH_PERCENTILE: 0.656,
  FIFTIETH_PERCENTILE: 0.846
};
const SWDMV4_RATINGS = {
  FIFTH_PERCENTILE: 0.04,
  TENTH_PERCENTILE: 0.079,
  TWENTIETH_PERCENTILE: 0.145,
  THIRTIETH_PERCENTILE: 0.209,
  FORTIETH_PERCENTILE: 0.278,
  FIFTIETH_PERCENTILE: 0.359
};
